File: ios/Sources/Scoring.swift
import Foundation

/// 축잘알 점수 엔진 — 기술 명세 2장. 웹 `src/lib/scoring.ts` 와 같은 공식이다.
///
/// 여기 계산은 "미리보기" 용도다. 실제 지수·포인트는 서버 정산이 진실의 원천이며,
/// 결과가 오면 서버 값으로 덮어쓴다 (명세 11.1).
/// 두 클라이언트가 다른 숫자를 보여주면 안 되므로 상수는 웹과 반드시 같아야 한다.
enum Outcome: String, CaseIterable, Codable {
    case home = "HOME", draw = "DRAW", away = "AWAY"
    var index: Int { Outcome.allCases.firstIndex(of: self)! }
}

enum Confidence: Int, CaseIterable, Codable {
    case hunch = 1, fairly = 2, certain = 3
    var label: String {
        switch self {
        case .hunch:   return "감이 와요"
        case .fairly:  return "꽤 확신해요"
        case .certain: return "이건 확실해요"
        }
    }
}

typealias Distribution = [Double]   // 항상 [HOME, DRAW, AWAY] 순서

enum Scoring {
    /// 로그 스코어 배율. 경제 파라미터가 아니라 실력 측정 단위라 함부로 바꾸지 않는다.
    static let K = 40.0
    static let clampMin = -60.0
    static let clampMax = 90.0
    static let participationPoints = 3

    /// 확신도 → 기준선에서 내 픽 쪽으로 이동하는 비율
    static let shift: [Confidence: Double] = [.hunch: 0.15, .fairly: 0.35, .certain: 0.6]

    /// 유저 확률분포. 확신도는 절대 확률이 아니라 기준선에서의 상대 이동이다.
    /// 이 설계 덕분에 "맞혔는데 점수가 깎이는" 경우가 생기지 않는다 (명세 2.3).
    static func userDistribution(_ q: Distribution, _ pick: Outcome, _ c: Confidence) -> Distribution {
        let i = pick.index
        var p = [0.0, 0.0, 0.0]
        p[i] = q[i] + shift[c]! * (1 - q[i])

        let rest = 1 - p[i]
        let otherSum = q.enumerated().reduce(0.0) { $1.offset == i ? $0 : $0 + $1.element }
        for j in 0..<3 where j != i { p[j] = rest * q[j] / otherSum }
        return p
    }

    /// Δ지수 = round(K · log2(p(실제) / q(실제))), clamp 적용 (명세 2.4)
    static func deltaRating(_ q: Distribution, _ pick: Outcome, _ c: Confidence, actual: Outcome) -> Int {
        let p = userDistribution(q, pick, c)
        let a = actual.index
        let raw = K * log2(p[a] / q[a])
        return Int(max(clampMin, min(clampMax, raw.rounded())))
    }

    /// 스트릭 보너스 = min(연속적중, 10) × 2 (명세 2.8)
    static func streakBonus(_ streak: Int) -> Int { min(max(streak, 0), 10) * 2 }

    struct Preview {
        let ifCorrect: Int
        let ifWrong: Int
        let pointsIfCorrect: Int
        let pointsIfWrong: Int
    }

    static func preview(_ q: Distribution, _ pick: Outcome, _ c: Confidence, streak: Int = 0) -> Preview {
        // 틀렸을 때 손해는 어떤 오답이든 같다 (명세 2.5) → 아무 오답이나 하나로 계산한다
        let wrong = Outcome.allCases[(pick.index + 1) % 3]
        let ifCorrect = deltaRating(q, pick, c, actual: pick)
        let ifWrong = deltaRating(q, pick, c, actual: wrong)
        return Preview(
            ifCorrect: ifCorrect,
            ifWrong: ifWrong,
            pointsIfCorrect: max(0, ifCorrect) + participationPoints + streakBonus(streak),
            pointsIfWrong: participationPoints)
    }
}

// MARK: - 레벨 · 티어

struct LevelState {
    let level: Int
    let into: Int
    let need: Int
    let progress: Double
}

enum Tier: String {
    case placement = "PLACEMENT", bronze = "BRONZE", silver = "SILVER", gold = "GOLD"
    case platinum = "PLATINUM", diamond = "DIAMOND", master = "MASTER", grandmaster = "GRANDMASTER"

    var label: String {
        switch self {
        case .placement: return "배치 중"
        case .bronze: return "브론즈"
        case .silver: return "실버"
        case .gold: return "골드"
        case .platinum: return "플래티넘"
        case .diamond: return "다이아몬드"
        case .master: return "마스터"
        case .grandmaster: return "그랜드마스터"
        }
    }
}

enum Progression {
    static let maxLevel = 50
    static let placementMatches = 20

    static func pointsForNextLevel(_ level: Int) -> Int { 100 + 25 * (level - 1) }

    /// 레벨 L 도달에 필요한 누적 포인트 = (L−1)(75 + 12.5L)
    static func cumulativePoints(_ level: Int) -> Int {
        Int((Double(level - 1) * (75 + 12.5 * Double(level))).rounded())
    }

    static func level(fromPoints points: Int) -> LevelState {
        var level = 1
        while level < maxLevel && cumulativePoints(level + 1) <= points { level += 1 }
        let base = cumulativePoints(level)
        let need = level >= maxLevel ? 0 : pointsForNextLevel(level)
        let into = points - base
        return LevelState(
            level: level, into: into, need: need,
            progress: need == 0 ? 1 : min(1, Double(into) / Double(need)))
    }

    /// 배치를 마쳤는데 아직 순위표에 오르지 않았으면 티어는 '없음'이다.
    /// 브론즈로 떨어뜨리면 실제로 받은 적 없는 등급을 보여주게 된다.
    static func tier(topPercent: Double?, settledMatches: Int) -> Tier? {
        if settledMatches < placementMatches { return .placement }
        guard let p = topPercent else { return nil }
        switch p {
        case ...1:  return .grandmaster
        case ...5:  return .master
        case ...15: return .diamond
        case ...30: return .platinum
        case ...50: return .gold
        case ...75: return .silver
        default:    return .bronze
        }
    }
}

File: src/lib/scoring.ts
/**
 * 축잘알 점수 엔진 — 기술 명세 2장의 공식을 그대로 옮긴 것.
 *
 * 이 파일의 계산은 "미리보기" 용도다. 실제 지수/포인트는 서버 정산이 진실의 원천이며,
 * 클라이언트 값은 결과가 오면 서버 값으로 덮어쓴다. (명세 11.1)
 */

export type Outcome = 'HOME' | 'DRAW' | 'AWAY';
export type Confidence = 1 | 2 | 3;

/** 결과 배열 인덱스 순서. 모든 확률 배열은 이 순서를 따른다. */
export const OUTCOMES: readonly Outcome[] = ['HOME', 'DRAW', 'AWAY'] as const;

/** 로그 스코어 배율. 경제 파라미터가 아니라 실력 측정 단위이므로 함부로 바꾸지 않는다. */
export const K = 40;
export const CLAMP_MIN = -60;
export const CLAMP_MAX = 90;

/** 확신도 → 기준선에서 내 픽 쪽으로 이동하는 비율 */
export const CONFIDENCE_SHIFT: Record<Confidence, number> = { 1: 0.15, 2: 0.35, 3: 0.6 };

export const CONFIDENCE_LABEL: Record<Confidence, string> = {
  1: '감이 와요',
  2: '꽤 확신해요',
  3: '이건 확실해요',
};

export type Distribution = [number, number, number];

export function outcomeIndex(o: Outcome): 0 | 1 | 2 {
  return OUTCOMES.indexOf(o) as 0 | 1 | 2;
}

/** 합이 1이 되도록 정규화하고 [0.03, 0.94]로 클램프한다. (명세 2.2) */
export function normalizeBaseline(raw: Distribution): Distribution {
  const clamped = raw.map((v) => Math.min(0.94, Math.max(0.03, v)));
  const sum = clamped.reduce((a, b) => a + b, 0);
  return clamped.map((v) => v / sum) as Distribution;
}

/**
 * 유저 확률분포. 확신도는 절대 확률이 아니라 기준선에서의 상대 이동으로 해석한다.
 * 이 설계 덕분에 "맞혔는데 점수가 깎이는" 경우가 생기지 않는다. (명세 2.3)
 */
export function userDistribution(
  baseline: Distribution,
  pick: Outcome,
  confidence: Confidence
): Distribution {
  const q = baseline;
  const i = outcomeIndex(pick);
  const p: number[] = [0, 0, 0];
  p[i] = q[i] + CONFIDENCE_SHIFT[confidence] * (1 - q[i]);

  const rest = 1 - p[i];
  const otherSum = q.reduce((acc, v, idx) => (idx === i ? acc : acc + v), 0);
  for (let j = 0; j < 3; j++) {
    if (j !== i) p[j] = (rest * q[j]) / otherSum;
  }
  return p as Distribution;
}

/** Δ지수 = round(K · log2(p(실제) / q(실제))), clamp 적용 (명세 2.4) */
export function deltaRating(
  baseline: Distribution,
  pick: Outcome,
  confidence: Confidence,
  actual: Outcome
): number {
  const p = userDistribution(baseline, pick, confidence);
  const a = outcomeIndex(actual);
  const raw = K * Math.log2(p[a] / baseline[a]);
  return Math.max(CLAMP_MIN, Math.min(CLAMP_MAX, Math.round(raw)));
}

export interface ScorePreview {
  /** 내가 픽한 결과에 부여한 확률 */
  pPick: number;
  /** 맞았을 때 지수 변동 (양수) */
  ifCorrect: number;
  /** 틀렸을 때 지수 변동 (음수) — 어떤 오답이든 동일하다 (명세 2.5) */
  ifWrong: number;
  /** 맞았을 때 적립 포인트 */
  pointsIfCorrect: number;
  /** 틀렸을 때 적립 포인트 (참여 보너스만) */
  pointsIfWrong: number;
}

export const PARTICIPATION_POINTS = 3;

/** 스트릭 보너스 = min(연속적중, 10) × 2 (명세 2.8) */
export function streakBonus(streak: number): number {
  return Math.min(Math.max(streak, 0), 10) * 2;
}

export function previewScore(
  baseline: Distribution,
  pick: Outcome,
  confidence: Confidence,
  streak = 0
): ScorePreview {
  const p = userDistribution(baseline, pick, confidence);
  const i = outcomeIndex(pick);
  const wrong = OUTCOMES[(i + 1) % 3];

  const ifCorrect = deltaRating(baseline, pick, confidence, pick);
  const ifWrong = deltaRating(baseline, pick, confidence, wrong);

  return {
    pPick: p[i],
    ifCorrect,
    ifWrong,
    pointsIfCorrect: Math.max(0, ifCorrect) + PARTICIPATION_POINTS + streakBonus(streak),
    pointsIfWrong: PARTICIPATION_POINTS,
  };
}

/* ------------------------------------------------------------------ */
/* 레벨 · 티어                                                          */
/* ------------------------------------------------------------------ */

/** 레벨 L → L+1 에 필요한 포인트 */
export function pointsForNextLevel(level: number): number {
  return 100 + 25 * (level - 1);
}

/** 레벨 L 도달에 필요한 누적 포인트 = (L−1)(75 + 12.5L) */
export function cumulativePointsForLevel(level: number): number {
  return Math.round((level - 1) * (75 + 12.5 * level));
}

export const MAX_LEVEL = 50;

export interface LevelState {
  level: number;
  /** 현재 레벨 안에서 쌓은 포인트 */
  into: number;
  /** 다음 레벨까지 필요한 총량 */
  need: number;
  /** 0..1 */
  progress: number;
}

export function levelFromPoints(lifetimePoints: number): LevelState {
  let level = 1;
  while (level < MAX_LEVEL && cumulativePointsForLevel(level + 1) <= lifetimePoints) {
    level++;
  }
  const base = cumulativePointsForLevel(level);
  const need = level >= MAX_LEVEL ? 0 : pointsForNextLevel(level);
  const into = lifetimePoints - base;
  return { level, into, need, progress: need === 0 ? 1 : Math.min(1, into / need) };
}

export type Tier =
  | 'PLACEMENT'
  | 'BRONZE'
  | 'SILVER'
  | 'GOLD'
  | 'PLATINUM'
  | 'DIAMOND'
  | 'MASTER'
  | 'GRANDMASTER';

export const TIER_LABEL: Record<Tier, string> = {
  PLACEMENT: '배치 중',
  BRONZE: '브론즈',
  SILVER: '실버',
  GOLD: '골드',
  PLATINUM: '플래티넘',
  DIAMOND: '다이아몬드',
  MASTER: '마스터',
  GRANDMASTER: '그랜드마스터',
};

export const PLACEMENT_MATCHES = 20;

/**
 * 백분위(상위 %)로 티어를 정한다. 유저가 늘어도 자동으로 균형이 맞는다. (명세 3.3)
 *
 * 배치를 마쳤는데 아직 순위표에 오르지 않았으면(= 다음 발표 전) 티어는 '없음'이다.
 * 이때 브론즈로 떨어뜨리면 실제로 받은 적 없는 등급을 보여주게 된다.
 */
export function tierFromPercentile(
  topPercent: number | null,
  settledMatches: number
): Tier | null {
  if (settledMatches < PLACEMENT_MATCHES) return 'PLACEMENT';
  if (topPercent == null) return null;
  if (topPercent <= 1) return 'GRANDMASTER';
  if (topPercent <= 5) return 'MASTER';
  if (topPercent <= 15) return 'DIAMOND';
  if (topPercent <= 30) return 'PLATINUM';
  if (topPercent <= 50) return 'GOLD';
  if (topPercent <= 75) return 'SILVER';
  return 'BRONZE';
}
